File: hilla_ride_ios/HelloTukTuk/Core/Services/WalletService.swift
import FirebaseAuth
import FirebaseFirestore
import FirebaseFunctions
import FirebaseStorage
import Foundation

enum WalletServiceError: LocalizedError {
    case unauthorized
    case invalidAmount
    case screenshotRequired
    case uploadFailed
    case invalidWithdrawal

    var errorDescription: String? {
        switch self {
        case .unauthorized: return "Sign in required."
        case .invalidAmount: return "Minimum recharge is 1000 IQD."
        case .screenshotRequired: return "Payment screenshot is required."
        case .uploadFailed: return "Could not upload screenshot."
        case .invalidWithdrawal: return "Invalid withdrawal request."
        }
    }
}

struct WalletWithdrawalSubmitResult: Equatable {
    let requestId: String
    let referenceId: String
}

final class WalletService {
    private let firestore = Firestore.firestore()
    private let storage = Storage.storage()
    private let functions = Functions.functions(region: "us-central1")
    private let auth = Auth.auth()

    func fetchConfig() async throws -> WalletConfig {
        let doc = try await firestore.collection("config").document("wallet").getDocument()
        return WalletConfig(data: doc.data())
    }

    func watchConfig() -> AsyncStream<WalletConfig> {
        AsyncStream { continuation in
            let listener = firestore.collection("config").document("wallet")
                .addSnapshotListener { snapshot, _ in
                    continuation.yield(WalletConfig(data: snapshot?.data()))
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func watchLedger(driverId: String) -> AsyncStream<[WalletLedgerEntry]> {
        AsyncStream { continuation in
            let listener = firestore.collection("walletLedger")
                .whereField("driverId", isEqualTo: driverId)
                .order(by: "createdAt", descending: true)
                .limit(to: 50)
                .addSnapshotListener { snapshot, _ in
                    let entries = snapshot?.documents.compactMap {
                        WalletLedgerEntry(documentID: $0.documentID, data: $0.data())
                    } ?? []
                    continuation.yield(entries)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func watchMyWithdrawals(driverId: String) -> AsyncStream<[WalletWithdrawalRequest]> {
        AsyncStream { continuation in
            let listener = firestore.collection("walletWithdrawalRequests")
                .whereField("driverId", isEqualTo: driverId)
                .order(by: "createdAt", descending: true)
                .limit(to: 40)
                .addSnapshotListener { snapshot, _ in
                    let items = snapshot?.documents.compactMap {
                        WalletWithdrawalRequest(documentID: $0.documentID, data: $0.data())
                    } ?? []
                    continuation.yield(items)
                }
            continuation.onTermination = { _ in listener.remove() }
        }
    }

    func submitWithdrawalRequest(
        amountIqd: Int,
        cardholderName: String,
        cardNumber: String
    ) async throws -> WalletWithdrawalSubmitResult {
        let digits = cardNumber.filter(\.isNumber)
        guard amountIqd > 0, !cardholderName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw WalletServiceError.invalidWithdrawal
        }
        let result = try await functions.httpsCallable("submitWalletWithdrawalRequest").call([
            "amountIqd": amountIqd,
            "cardholderName": cardholderName.trimmingCharacters(in: .whitespacesAndNewlines),
            "cardNumber": digits
        ])
        let data = result.data as? [String: Any] ?? [:]
        return WalletWithdrawalSubmitResult(
            requestId: data["requestId"] as? String ?? "",
            referenceId: data["referenceId"] as? String ?? ""
        )
    }

    func cancelWithdrawalRequest(requestId: String) async throws {
        guard !requestId.isEmpty else { throw WalletServiceError.invalidWithdrawal }
        _ = try await functions.httpsCallable("cancelWalletWithdrawalRequest").call([
            "requestId": requestId
        ])
    }

    func uploadRechargeScreenshot(driverId: String, data: Data) async throws -> String {
        guard let uid = auth.currentUser?.uid, uid == driverId else {
            throw WalletServiceError.unauthorized
        }
        guard !data.isEmpty else { throw WalletServiceError.uploadFailed }
        let id = UUID().uuidString
        let ref = storage.reference().child("wallet_recharges/\(driverId)/\(id).jpg")
        let meta = StorageMetadata()
        meta.contentType = "image/jpeg"
        _ = try await ref.putDataAsync(data, metadata: meta)
        return try await ref.downloadURL().absoluteString
    }

    func submitRechargeRequest(
        amountIqd: Int,
        method: String,
        screenshotUrl: String,
        referenceNumber: String,
        notes: String
    ) async throws {
        guard amountIqd >= 1000 else { throw WalletServiceError.invalidAmount }
        guard !screenshotUrl.isEmpty else { throw WalletServiceError.screenshotRequired }
        _ = try await functions.httpsCallable("submitWalletRechargeRequest").call([
            "amountIqd": amountIqd,
            "method": method,
            "screenshotUrl": screenshotUrl,
            "referenceNumber": referenceNumber,
            "notes": notes
        ])
    }
}

File: lib/core/services/wallet_service.dart
import 'dart:typed_data';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cloud_functions/cloud_functions.dart';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:hilla_ride/core/models/wallet_models.dart';
import 'package:uuid/uuid.dart';

class WalletService {
  WalletService({
    FirebaseFirestore? firestore,
    FirebaseFunctions? functions,
    FirebaseStorage? storage,
  })  : _firestore = firestore ?? FirebaseFirestore.instance,
        _functions = functions ??
            FirebaseFunctions.instanceFor(region: 'us-central1'),
        _storage = storage ?? FirebaseStorage.instance;

  final FirebaseFirestore _firestore;
  final FirebaseFunctions _functions;
  final FirebaseStorage _storage;

  Stream<WalletConfig> watchConfig() {
    return _firestore.collection('config').doc('wallet').snapshots().map(
          (doc) => WalletConfig.fromMap(doc.data()),
        );
  }

  Future<WalletConfig> fetchConfig() async {
    final doc = await _firestore.collection('config').doc('wallet').get();
    return WalletConfig.fromMap(doc.data());
  }

  Stream<List<WalletLedgerEntry>> watchLedger(String driverId, {int limit = 50}) {
    return _firestore
        .collection('walletLedger')
        .where('driverId', isEqualTo: driverId)
        .orderBy('createdAt', descending: true)
        .limit(limit)
        .snapshots()
        .map(
          (snap) => snap.docs.map(WalletLedgerEntry.fromDoc).toList(),
        );
  }

  Stream<List<WalletRechargeRequest>> watchMyRechargeRequests(String driverId) {
    return _firestore
        .collection('walletRechargeRequests')
        .where('driverId', isEqualTo: driverId)
        .orderBy('createdAt', descending: true)
        .limit(40)
        .snapshots()
        .map(
          (snap) => snap.docs.map(WalletRechargeRequest.fromDoc).toList(),
        );
  }

  Stream<List<WalletRechargeRequest>> watchPendingRechargeRequests() {
    return _firestore
        .collection('walletRechargeRequests')
        .where('status', isEqualTo: WalletRechargeStatus.pending.value)
        .orderBy('createdAt', descending: true)
        .limit(100)
        .snapshots()
        .map(
          (snap) => snap.docs.map(WalletRechargeRequest.fromDoc).toList(),
        );
  }

  Future<String> uploadRechargeScreenshot({
    required String driverId,
    required Uint8List bytes,
    String contentType = 'image/jpeg',
  }) async {
    final id = const Uuid().v4();
    final ref = _storage
        .ref()
        .child('wallet_recharges')
        .child(driverId)
        .child('$id.jpg');
    await ref.putData(bytes, SettableMetadata(contentType: contentType));
    return ref.getDownloadURL();
  }

  Future<void> submitRechargeRequest({
    required int amountIqd,
    required WalletRechargeMethod method,
    required String screenshotUrl,
    String referenceNumber = '',
    String notes = '',
  }) async {
    await _functions.httpsCallable('submitWalletRechargeRequest').call({
      'amountIqd': amountIqd,
      'method': method.value,
      'screenshotUrl': screenshotUrl,
      'referenceNumber': referenceNumber.trim(),
      'notes': notes.trim(),
    });
  }

  Stream<List<WalletLedgerEntry>> watchRecentLedger({int limit = 500}) {
    return _firestore
        .collection('walletLedger')
        .orderBy('createdAt', descending: true)
        .limit(limit)
        .snapshots()
        .map(
          (snap) => snap.docs.map(WalletLedgerEntry.fromDoc).toList(),
        );
  }

  Future<void> reviewRechargeRequest({
    required String requestId,
    required bool approve,
    String rejectionReason = '',
    int? approvedAmountIqd,
  }) async {
    await _functions.httpsCallable('reviewWalletRechargeRequest').call({
      'requestId': requestId,
      'approve': approve,
      'rejectionReason': rejectionReason.trim(),
      if (approvedAmountIqd != null) 'approvedAmountIqd': approvedAmountIqd,
    });
  }

  Future<void> adjustWallet({
    required String driverId,
    required int amountIqd,
    required String note,
  }) async {
    await _functions.httpsCallable('adjustDriverWallet').call({
      'driverId': driverId,
      'amountIqd': amountIqd,
      'note': note.trim(),
    });
  }

  Future<void> saveConfig(WalletConfig config) async {
    await _functions.httpsCallable('saveWalletConfig').call(config.toMap());
  }

  /// Backfill `walletBalanceIqd` / `walletStatus` on drivers missing them.
  Future<({int updated, int total})> ensureDriverWallets() async {
    final result =
        await _functions.httpsCallable('ensureDriverWallets').call();
    final data = Map<String, dynamic>.from(result.data as Map? ?? {});
    return (
      updated: (data['updated'] as num?)?.toInt() ?? 0,
      total: (data['total'] as num?)?.toInt() ?? 0,
    );
  }
}
